### ISSUE TITLE
Strange mesh in testing

### ISSUE THREAD

**Issue by dldaisy** [Jeff-sjtu/HybrIK#50]
Hi, I run testing on wild images and found sometimes it will output extremely unnatural body mesh.
Here is the input image:
![000003](https://user-images.githubusercontent.com/37493043/133054467-6fb5d3b2-6a7d-4668-8fa0-a52b2f73fd98.png)
And below are the screenshots of the reconstructed image from front and side view:
<img width="367" alt="Screen Shot 2021-09-13 at 4 51 51 PM" src="https://user-images.githubusercontent.com/37493043/133054872-bed48bc5-8782-48db-9409-876502fdc9d1.png">
<img width="478" alt="Screen Shot 2021-09-13 at 4 58 38 PM" src="https://user-images.githubusercontent.com/37493043/133055229-4491a013-5cc7-46aa-b52a-c9d11e1755a3.png">
As you can see, the body seemed to be twisted 180 degree near the joint SPINE_3, as the result, the upper body is inside-out.
I also test the model on several other images and find that this problem would very much likely to occur when the input images don't include full body(for example, only partial body or half body).
May I ask why this weird problem exists? Thanks a lot!

**Comment by dldaisy** [comment]
Got it, Thank you!

**Comment by puppet101** [comment]
@dldaisy can you share your testing code for in the wild images?

**Comment by coollx** [comment]
@dldaisy After browsing all the issues and posts I still have no idea to generating the test code. Do you mind to share your own testing code for the wild images? This would help a lot. Thanks in advance!

**Comment by biansy000** [comment]
hello @dldaisy , we found a bug in calculating the rotation matrix for spine3 in inference, and the updated code is uploaded in ``fix_wrong_rotmat`` branch. The bug is probably the cause of the abnormal twist.

**Comment by dldaisy** [comment]
@biansy000 Thanks for the update. I think the bug is due to the matrix recovered from svd is not guaranteed to have positive determination right? Which means if the determination is negative, the matrix will embed a reflection operation in it instead of being a real rotation matrix.
The great code piece updated definitely solve the problem by modifying the matrix to a rotation matrix. But I still have a question: after this modification, would the optimality of the rotation matrix got from svd still be guaranteed, since it put extra constraint to your proof in the supplementary material?
Thanks a lot.

**Comment by biansy000** [comment]
@dldaisy Yes, the bug is due to the negative determinant of matrix, and you can refer to [ref1](http://www.kwon3d.com/theory/jkinem/rotmat.html) or [ref2](https://www.ece.uvic.ca/~bctill/papers/mocap/Challis_1995.pdf) for further infomation. But neither of the references make a detailed proof.

**Comment by dldaisy** [comment]
I see. Thanks a lot.
